src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1144,7 +1131,8 @@ private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials
                         SR.net_http_client_execution_error,
                         WinHttpException.CreateExceptionUsingLastError());
                 }
-
+                requestHandle.SetParentHandle(connectHandle);
```

Comment:
Shouldn't the hierarchy include the connectionHandle as well? [session->connection->request](https://msdn.microsoft.com/en-us/library/windows/desktop/aa383880(v=vs.85).aspx#_hierarchy)

I think having a flat architecture (similar to the Strategy pattern that was originally hinted by the extraction of the context into the `state` object) would be much simpler to understand.

I've based the design in ClientWebSockets on your original state object and, since managed references are kept towards the state, it's guaranteed not to release the safehandles until all requests are done.